Shared generics opcodes

davidwrighton opened the pull request:
Implement the vast majority of the opcodes that are affected by shared generics
- box, unbox, unbox.any, castclass, isinst, ldtoken, newarr, newobj of an mdarray

Add logic to enhance pretty printing of the Interp opcodes for these things. We can now see strings, and various magic pointer values get displayed in the disasm outputs.

Handle the 0 indirections generic lookup scenario

Many opcodes were able to use a set of generic helpercall opcodes. These opcodes have a naming convention of
- `a` stands for addr of dvar or svar
- `g` stands for do a generic lookup
- `p` stands for pointer (acquired from either an svar or a dataitem)
- `v` stands for a void return.

Its not fully descriptive, but naming is hard, and this works.

davidwrighton committed:
Implement most of the shared generics opcodes needed
- boxint, unbox, unbox.any, cast, isinst, ldtoken, newarr, newobj

Added additional diagnostic printing capabilities, to print strings, and method/class names in interesting scenarios

Known continuing work
- MDArray new
- static variables and class constructor triggering

davidwrighton committed:
Add multidimensional array handling

davidwrighton committed:
Merge branch 'main' of github.com:dotnet/runtime into shared_generics_opcodes_fillout

davidwrighton committed:
Handle the generic shared 0 indirections scenario

davidwrighton committed:
Make helper call opcodes have some semblance of a naming convention

kg commented on src/coreclr/interpreter/compiler.h:
```diff
@@ -403,6 +468,35 @@ class InterpCompiler
+        GenericHandleData() = default;
+
+        HelperArgType argType = HelperArgType::Value;
+        int genericVar = -1; // Set to a meaningful value if HelperArgType is GenericResolution
```
Something more specific than 'a meaningful value' would be helpful here since by browsing around the vicinity I can't tell what makes the value meaningful or how I would distinguish between a valid genericVar and an invalid one. It may become obvious to me as I continue reading the PR, though!

My assumption upon first glance is it's the zero-based index of the generic parameter in the method's set of reachable generic parameters (where i use 'reachable generic parameters' to mean the set of all the enclosing type parameters + all the method parameters)

kg replied:
OK, I'm at this point again and it's 'the interp var containing the generic dictionary' I think?

kg commented on src/coreclr/interpreter/compiler.cpp:
```diff
@@ -2206,26 +2235,189 @@ InterpCompiler::InterpEmbedGenericResult InterpCompiler::EmitGenericHandle(CORIN
+        m_pLastNewIns->SetSVar(arg2);
+        m_pLastNewIns->SetDVar(resultVar);
+
+        m_pLastNewIns->data[1] = handleData.dataItemIndex;
```
[nitpick] There's an asymmetry here where in the GenericResolution arm of these ifs, we set data[0] and data[1] right next to each other, but in the other arm we set them separately. It seems like it could hide mistakes. It would be cool if both sides were the same shape so the differences were clearer to the reader.

davidwrighton replied:
I've made this cleaner.

kg commented on src/coreclr/interpreter/compiler.cpp:
```diff
@@ -2206,26 +2235,189 @@ InterpCompiler::InterpEmbedGenericResult InterpCompiler::EmitGenericHandle(CORIN
+    if (embedInfo.lookup.lookupKind.needsRuntimeLookup)
+    {
+        CORINFO_RUNTIME_LOOKUP_KIND runtimeLookupKind = embedInfo.lookup.lookupKind.runtimeLookupKind;
+        InterpGenericLookup *pRuntimeLookup = (InterpGenericLookup*)AllocMethodData(sizeof(InterpGenericLookup));
```
Is it the case that GenericHandleToGenericHandleData could get called for the same embedInfo multiple times, and we'd bloat the dataitems table with lots of duplicate lookups? i.e. if a method has two generic parameters T and U, are we going to end up with a unique dataitem for every use of T and every use of U? Or just two data items, one for T and one for U, that get reused through some other mechanism I haven't noticed yet?

davidwrighton replied:
Yeah, this is a problem. I've added a general purpose hashing mechanism for embedding data into the dataitems array, and swapped over to using that hash even for regular pointers. This should address any reasonable scalability issues here.

kg commented on src/coreclr/interpreter/compiler.cpp:
```diff
@@ -2563,26 +2749,42 @@ void InterpCompiler::EmitCall(CORINFO_RESOLVED_TOKEN* pConstrainedToken, bool re
+                else
+                {
+                    AddIns(INTOP_NEWMDARR);
+                    DeclarePointerIsClass(resolvedCallToken.hClass);
```
Missing a SetSVars?

davidwrighton replied:
No, the call path is convoluted and weird, and the SetSVar happens... after the switch is over. Its messy enough that we may want to clean this up, but this is following existing practices.

kg commented on src/coreclr/interpreter/compiler.cpp:
```diff
@@ -2563,26 +2749,42 @@ void InterpCompiler::EmitCall(CORINFO_RESOLVED_TOKEN* pConstrainedToken, bool re
                         // Normal newobj call
                         AddIns(INTOP_NEWOBJ);
-                        m_pLastNewIns->data[1] = newObjType.dataItemIndex;
+                        m_pLastNewIns->data[1] = newObjData.dataItemIndex;
```
Missing a SetSVar? Is it implicit?

davidwrighton replied:
Yes.

kg commented on src/coreclr/vm/interpexec.cpp:
```diff
@@ -291,6 +293,67 @@ template <typename TResult, typename TSource> void ConvOvfHelper(int8_t *stack,
     }
 }
 
+#define DO_GENERIC_LOOKUP(genericVar, lookupIndex) \
```
I agree with copilot that I'd love to see this outlined into function(s). As-is the line number information when debugging it is gonna be kinda miserable.

janvorli replied:
I agree, I was debugging an issue with it in the past and I had to expand the macro manually to be able to figure out what's wrong.

davidwrighton replied:
K, I've cleaned that up.

kg commented on src/coreclr/vm/interpexec.cpp:
```diff
@@ -291,6 +293,67 @@ template <typename TResult, typename TSource> void ConvOvfHelper(int8_t *stack,
+                    if (pLookup->indirections != 0)                                                         \
+                    {                                                                                       \
+                        lookup = *(uint8_t**)(lookup + pLookup->offsets[0]);                                \
+                        if (pLookup->indirections >= 3)                                                     \
```
This and the 4 look like they're off by 1, but I assume that's on purpose and goes with the 'last indirection is the size' below

kg commented on src/coreclr/vm/interpexec.cpp:
```diff
@@ -291,6 +293,67 @@ template <typename TResult, typename TSource> void ConvOvfHelper(int8_t *stack,
+                        pMD = LOCAL_VAR(genericVar, MethodDesc*);                                           \
+                        lookup = (uint8_t*)pMD;                                                             \
+                    }                                                                                       \
+                    void* result = 0;                                                                       \
```
[nitpick] It'd be nice if this name made it a little clearer that it's from the macro, or if it was defined at the top of the macro instead of hiding in the middle

kg commented on src/coreclr/interpreter/compiler.h:
```diff
@@ -348,10 +348,75 @@ class InterpCompiler
+    TArray<PointerToName> m_pointerToNameMap;
+    bool PointerInNameMap(void* ptr)
+    {
+        for (int32_t i = 0; i < m_pointerToNameMap.GetSize(); i++)
```
fyi if you look at dn-simdhash-ght-compatible.c you can see how to define custom callbacks and whatnot. but there's nothing wrong with using ptr_ptr if you're okay with an extra allocation per item.

davidwrighton committed:
First round of PR feedback

davidwrighton commented on src/coreclr/interpreter/compiler.cpp:
```diff
@@ -2206,26 +2235,189 @@ InterpCompiler::InterpEmbedGenericResult InterpCompiler::EmitGenericHandle(CORIN
+    }
+}
+
+void InterpCompiler::EmitPushUnboxAnyNullable(const CORINFO_GENERICHANDLE_RESULT& arg1, int arg2, StackType resultStackType, CORINFO_CLASS_HANDLE clsHndStack)
```
The idea is to allow sharing code with various paths that use things like EmitPushCallHelper which obviously can't pull a random thing off of the evaluation stack.

davidwrighton commented on src/coreclr/interpreter/compiler.cpp:
```diff
@@ -4774,34 +4978,87 @@ void InterpCompiler::GenerateCode(CORINFO_METHOD_INFO* methodInfo)
-                if (m_compHnd->isValueClass(clsHnd))
+                CORINFO_RESOLVED_TOKEN resolvedToken;
+                ResolveToken(token, CORINFO_TOKENKIND_Box, &resolvedToken);
+                if (m_compHnd->isValueClass(resolvedToken.hClass))
```
That can't happen in CoreCLR style shared generics.

davidwrighton commented on src/coreclr/interpreter/compiler.cpp:
```diff
@@ -4774,34 +4978,87 @@ void InterpCompiler::GenerateCode(CORINFO_METHOD_INFO* methodInfo)
+                m_compHnd->embedGenericHandle(&resolvedToken, false, m_methodInfo->ftn, &embedInfo);
+                DeclarePointerIsClass((CORINFO_CLASS_HANDLE)embedInfo.compileTimeHandle);
+
+                m_pStackPointer--;
```
I've merged them to the extent that makes sense to me.

davidwrighton commented on src/coreclr/interpreter/compiler.cpp:
```diff
@@ -5187,23 +5452,18 @@ void InterpCompiler::GenerateCode(CORINFO_METHOD_INFO* methodInfo)
-                    PushInterpType(InterpTypeI, NULL);
-                m_pLastNewIns->SetDVar(m_pStackPointer[-1].var);
+                DeclarePointerIsClass((CORINFO_CLASS_HANDLE)embedInfo.compileTimeHandle);
+                EmitPushHelperCall_2(castingHelper, embedInfo, m_pStackPointer[0].var, g_stackTypeFromInterpType[*m_ip == CEE_CASTCLASS ? InterpTypeO : InterpTypeI], NULL);
```
Its ok. The clsHndStack is only actually meaningful when the stacktype is StackTypeVT, which. At other times it is sometimes filled with a type, and sometimes not. The presence of the type cannot be relied upon, but one could imagine using it when debugging or something if it happens to be there. We would only need to actually track these things for non-structs if we implemented the verifier logic, which is an entirely different problem than what we're solving today.

davidwrighton committed:
Use a map to avoid repeated allocations of things for the generic lookups, and embed the InterpGenericLookup in the dataitems array directly
Fix generic lookups to handle the usehelper case
Shrink the InterpGenericLookup to be the actual size needed, using smaller data fields when possible
Improve comments on the interpreter around what the sizeOffset means

davidwrighton committed:
Fix unix build breaks

davidwrighton committed:
Fix build issues

am11 commented on src/native/containers/dn-simdhash-string-ptr.h:
```diff
@@ -1,3 +1,7 @@
+#ifdef __cplusplus
```
```suggestion
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.

#ifdef __cplusplus
```

davidwrighton committed:
Update src/native/containers/dn-simdhash-string-ptr.h

Co-authored-by: Adeel Mujahid <3840695+am11@users.noreply.github.com>

kg commented on src/coreclr/interpreter/compiler.cpp:
```diff
@@ -2206,27 +2292,190 @@ InterpCompiler::InterpEmbedGenericResult InterpCompiler::EmitGenericHandle(CORIN
+        m_pLastNewIns->data[0] = GetDataItemIndexForHelperFtn(ftn);
+        m_pLastNewIns->data[1] = handleData.dataItemIndex;
+
+        m_pLastNewIns->SetDVar(resultVar);
```
At least one call site I found (CEE_LDTOKEN) does EmitPushHelperCall without calling SetSVar at the end, so are we missing one here?

kg replied:
Ah, I see, this particular opcode has no svar, it looks like.

kg commented on src/coreclr/interpreter/compiler.h:
```diff
@@ -29,6 +76,104 @@ INTERPRETER_NORETURN void NO_WAY(const char* message);
+
+        size_t sizeOfStruct = sizeof(InterpGenericLookup);
+
+        assert(sizeOfFieldsConcatenated == sizeOfStruct); // Assert that there is no padding in the struct, so a fixed size hash unaware of padding is safe to use
```
This can be a static_assert right?

davidwrighton replied:
But there is a good point that that doesn't really happen in release builds.

kg commented on src/coreclr/interpreter/compiler.h:
```diff
@@ -348,10 +493,67 @@ class InterpCompiler
+    const char* PointerIsStringLiteral = (const char*)0x3;
+
+    dn_simdhash_ptr_ptr_holder m_pointerToNameMap;
+    bool PointerInNameMap(void* ptr)
+    {
+        void* result;
+        if (dn_simdhash_ptr_ptr_try_get_value(m_pointerToNameMap.GetValue(), ptr, &result))
+        {
+            return true;
+        }
+        return false;
+    }
+    void AddPointerToNameMap(void* ptr, const char* name)
+    {
+        if (!PointerInNameMap(ptr))
+        {
+            dn_simdhash_ptr_ptr_try_add(m_pointerToNameMap.GetValue(), ptr, (void*)name);
+        }
+    }
```
As things are written right now, we don't need `PointerInNameMap` at all, and the `try_add` operation will harmlessly return 0 if the key `ptr` already exists.

kg commented on src/coreclr/interpreter/compiler.h:
```diff
@@ -348,10 +493,67 @@ class InterpCompiler
+    bool PointerInNameMap(void* ptr)
+    {
+        void* result;
+        if (dn_simdhash_ptr_ptr_try_get_value(m_pointerToNameMap.GetValue(), ptr, &result))
```
fyi, you can pass NULL instead of &result and simdhash try_get_value will do what you want (not copy, just return whether a value was found or not)

kg commented on src/coreclr/interpreter/compiler.h:
```diff
@@ -586,4 +835,41 @@ class InterpCompiler
+    }
+
+    // Assert that there is no padding in the struct, so a fixed size hash unaware of padding is safe to use
+    assert(sizeof(VarSizedData) == sizeof(void*));
+    assert(sizeof(T) % sizeof(void*) == 0);
+    assert(sizeof(VarSizedDataWithPayload<T>) == sizeof(T) + sizeof(void*));
```
These should probably also be static_assert instead of assert, so that they will be checked in release builds

BrzVlad commented on src/coreclr/vm/interpexec.cpp:
```diff
@@ -291,6 +293,79 @@ template <typename TResult, typename TSource> void ConvOvfHelper(int8_t *stack,
+    return result;
+}
+
+#define DO_GENERIC_LOOKUP(genericVar, lookupIndex) \
```
Do we even need this define anymore ? Doesn't seem to add more clarity. We could just replace it with a method returning result explicitly

davidwrighton commented on src/coreclr/vm/interpexec.cpp:
```diff
@@ -1533,9 +1608,81 @@ void InterpExecMethod(InterpreterFrame *pInterpreterFrame, InterpMethodContextFr
+                    int base = 4;
+                    DO_GENERIC_LOOKUP(ip[2], base + 1);
+
+                    HELPER_FTN_PP helperFtn = GetPossiblyIndirectHelper<HELPER_FTN_PP>(pMethod->pDataItems[ip[base]]);
```
Mmm... fair. I was just copying/pasting here, and mostly spending my time thinking about  the compiler side. I'm going to eliminate the code sharing between opcodes here, as I think its obfuscating the logic, and simplify this out.
